pyro/contrib/funsor/handlers/named_messenger.py

```diff
@@ -178,3 +205,211 @@ def __enter__(self):
+                    init if isinstance(i, int) and i < 1 else trans[x_prev]),
+
+                pyro.sample("y_{}".format(i),
+                            dist.Normal(Vindex(locs)[..., x_curr], 1.),
+                            obs=data[i])
+                x_prev = x_curr
+
+        #  trace.nodes["time"]["infer"]["step"]
+        #  frozenset({('x_0', 'x_tensor([0, 1])', 'x_tensor([1, 2])')})
+        #
+        #  pyro.vectorized_markov trace
+        #  ...
+        #  Sample Sites:
+        #      locs dist               | 3
+        #          value               | 3
+        #       log_prob               |
+        #       x_0 dist               |
+        #          value     3 1 1 1 1 |
+        #       log_prob     3 1 1 1 1 |
+        #       y_0 dist     3 1 1 1 1 |
+        #          value               |
+        #       log_prob     3 1 1 1 1 |
+        #  x_tensor([1, 2]) dist   3 1 1 1 1 2 |
+        #          value 3 1 1 1 1 1 1 |
+        #       log_prob 3 3 1 1 1 1 2 |
+        #  y_tensor([1, 2]) dist 3 1 1 1 1 1 2 |
+        #          value             2 |
+        #       log_prob 3 1 1 1 1 1 2 |
+        #
+        #  pyro.markov trace
+        #  ...
+        #  Sample Sites:
+        #      locs dist             | 3
+        #          value             | 3
+        #       log_prob             |
+        #       x_0 dist             |
+        #          value   3 1 1 1 1 |
+        #       log_prob   3 1 1 1 1 |
+        #       y_0 dist   3 1 1 1 1 |
+        #          value             |
+        #       log_prob   3 1 1 1 1 |
+        #       x_1 dist   3 1 1 1 1 |
+        #          value 3 1 1 1 1 1 |
+        #       log_prob 3 3 1 1 1 1 |
+        #       y_1 dist 3 1 1 1 1 1 |
+        #          value             |
+        #       log_prob 3 1 1 1 1 1 |
+        #       x_2 dist 3 1 1 1 1 1 |
+        #          value   3 1 1 1 1 |
+        #       log_prob 3 3 1 1 1 1 |
+        #       y_2 dist   3 1 1 1 1 |
+        #          value             |
+        #       log_prob   3 1 1 1 1 |
+
+    .. warning::  This is only correct if there is only one Markov
+        dimension per branch.
+
+    :param str name: A unique name of a Markov dimension to help inference algorithm
+        eliminate variables in the Markov chain.
+    :param int size: Length (size) of the Markov chain.
+    :param int dim: An optional dimension to use for this independence index.
+        If specified, ``dim`` should be negative, i.e. should index from the
+        right. If not specified, ``dim`` is set to the rightmost dim that is
+        left of all enclosing ``plate`` contexts.
+    :param int history: Memory (order) of the Markov chain. Also the number
+        of previous contexts visible from the current context. Defaults to 1.
+        If zero, this is similar to :class:`pyro.plate`.
+    :param bool keep: If true, frames are replayable. This is important
```

Comment:
Can you get rid of `keep` and all associated code for now? It's not clear that `keep=True` is compatible with vectorized enumeration. We can revisit this in future PRs if necessary.